[8.6][8.7] Investigate failing test ExporterDisableTest.exporterStaysDisabledAfterRestart

### <!-- Component -->

<!-- Zeebe- -->

### <!-- Affected version -->

<!-- -8.6 -->, <!-- -8.7 -->

### <!-- Severity -->

<!-- Unknown- -->

### <!-- Likelihood -->

<!-- High_ -->

### Description

[Slack discussion](https://camunda.slack.com/archives/C08MRKHJ0CD/p1755716456899509)
`ExporterDisableTest.exporterStaysDisabledAfterRestart` is frequently failing in stable/8.6 and stable/8.7 PRs and merge queues, there are also [flakiness](https://camunda.slack.com/archives/C071KP5BTHB/p1755756693781149) reported in `main`

### Steps to reproduce

In local, the test is consistently failing
https://github.com/camunda/camunda/actions/runs/17067732786/job/48388912890

### Current behavior

```
Error:  io.camunda.zeebe.it.exporter.ExporterDisableTest.exporterStaysDisabledAfterRestart -- Time elapsed: 198.6 s <<< ERROR!
java.util.concurrent.TimeoutException: exporterStaysDisabledAfterRestart() timed out after 120 seconds
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
	Suppressed: org.awaitility.core.ConditionTimeoutException: Condition with alias 'until cluster topology is complete' didn't complete within 3 minutes because assertion condition defined as a Lambda expression in io.camunda.zeebe.qa.util.cluster.TestCluster 
Expecting all elements of:
  [TestStandaloneGateway{nodeId = gateway-0}]
to satisfy given requirements, but these elements did not:

TestStandaloneGateway{nodeId = gateway-0}
error: Expected all partitions to be healthy, but partition <2> of broker <1> is <UNHEALTHY>.
		at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:167)
		at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
		at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
		at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:1006)
		at org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:790)
		at io.camunda.zeebe.qa.util.cluster.TestCluster.awaitCompleteTopology(TestCluster.java:316)
		at io.camunda.zeebe.qa.util.cluster.TestCluster.awaitCompleteTopology(TestCluster.java:294)
		at io.camunda.zeebe.qa.util.cluster.TestCluster.awaitCompleteTopology(TestCluster.java:280)
		at io.camunda.zeebe.it.exporter.ExporterDisableTest.exporterStaysDisabledAfterRestart(ExporterDisableTest.java:114)
		at java.base/java.lang.reflect.Method.invoke(Method.java:580)
		... 5 more
	Caused by: java.lang.AssertionError: 
```

### Expected behavior

Test to pass

### Environment

SM

### Version

- Component Version: [e.g. 2.0.0]


### Rootcause

According to the job history, it seems that it started to fail since the merge of https://github.com/camunda/camunda/pull/36847, I reverted this PR in my local, and the test is now passing

### Solution ideas

_No response_

### Dev -> QA handover

- Resources:
- Versions to validate:
- Release version (in which version this feature will be released):


### Links

_No response_

--- a/zeebe/broker/spotbugs/spotbugs-exclude.xml
+++ b/zeebe/broker/spotbugs/spotbugs-exclude.xml
@@ -16,5 +16,10 @@
     <Class name="io.camunda.zeebe.broker.partitioning.scaling.snapshot.SnapshotChunkRecord"/>
     <Bug pattern="SING_SINGLETON_HAS_NONPRIVATE_CONSTRUCTOR"/>
   </Match>
+  <Match>
+    <!-- This is a false positive for property 'commitPosition' because it is updated with in an actor. So atomic updates are guaranteed. -->
+    <Class name="io.camunda.zeebe.broker.system.partitions.impl.AsyncSnapshotDirector"/>
+    <Bug pattern="AT_NONATOMIC_OPERATIONS_ON_SHARED_VARIABLE"/>
+  </Match>
   <Bug pattern="AT_NONATOMIC_64BIT_PRIMITIVE"/>
 </FindBugsFilter>

--- a/zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java
+++ b/zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java
@@ -370,7 +370,7 @@ public final class AsyncSnapshotDirector extends Actor
   public void newPositionCommitted(final long currentCommitPosition) {
     actor.run(
         () -> {
-          commitPosition = currentCommitPosition;
+          commitPosition = Math.max(currentCommitPosition, commitPosition);
           final var futuresToComplete = commitAwaiters.headMap(commitPosition, true);
           futuresToComplete.forEach((k, f) -> f.complete(null));
           futuresToComplete.clear();

--- a/zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java
+++ b/zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java
@@ -396,8 +396,8 @@ public final class AsyncSnapshotDirector extends Actor
   }
 
   @VisibleForTesting
-  public long getCommitPosition() {
-    return commitPosition;
+  public ActorFuture<Long> getCommitPosition() {
+    return actor.call(() -> commitPosition);
   }
 
   private static final class InProgressSnapshot {

--- a/zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java
+++ b/zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java
@@ -22,6 +22,7 @@ import io.camunda.zeebe.snapshots.TransientSnapshot;
 import io.camunda.zeebe.snapshots.transfer.SnapshotTransferService;
 import io.camunda.zeebe.stream.impl.StreamProcessor;
 import io.camunda.zeebe.stream.impl.StreamProcessorMode;
+import io.camunda.zeebe.util.VisibleForTesting;
 import io.camunda.zeebe.util.health.FailureListener;
 import io.camunda.zeebe.util.health.HealthMonitorable;
 import io.camunda.zeebe.util.health.HealthReport;
@@ -392,6 +393,11 @@ public final class AsyncSnapshotDirector extends Actor
     final var result = actor.<PersistedSnapshot>createFuture();
     actor.run(() -> snapshot(inProgressSnapshot, true).onComplete(result));
     return result;
+  }
+
+  @VisibleForTesting
+  public long getCommitPosition() {
+    return commitPosition;
   }
 
   private static final class InProgressSnapshot {

--- a/zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/steps/SnapshotDirectorPartitionTransitionStep.java
+++ b/zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/steps/SnapshotDirectorPartitionTransitionStep.java
@@ -11,6 +11,7 @@ import io.atomix.raft.RaftServer.Role;
 import io.camunda.zeebe.broker.system.partitions.PartitionTransitionContext;
 import io.camunda.zeebe.broker.system.partitions.PartitionTransitionStep;
 import io.camunda.zeebe.broker.system.partitions.impl.AsyncSnapshotDirector;
+import io.camunda.zeebe.logstreams.log.LogStreamReader;
 import io.camunda.zeebe.scheduler.SchedulingHints;
 import io.camunda.zeebe.scheduler.future.ActorFuture;
 import io.camunda.zeebe.scheduler.future.CompletableActorFuture;
@@ -73,6 +74,18 @@ public final class SnapshotDirectorPartitionTransitionStep implements PartitionT
               context.getComponentHealthMonitor().registerComponent(director);
               if (targetRole == Role.LEADER) {
                 server.addCommittedEntryListener(director);
+
+                // Raft server will only notify if there is a new entry commited. If the node has
+                // just restarted or transitioned to leader, but there are no new processing records
+                // written or committed, the listener is not triggered. As a result the
+                // commitPosition in the snapshotDirector remain 0, thus preventing snapshots even
+                // if the state has changed after replaying previously committed events. Hence, set
+                // the commit position from the last record in the log stream.
+                try (final LogStreamReader logStreamReader =
+                    context.getLogStream().newLogStreamReader()) {
+                  final var commitPosition = logStreamReader.seekToEnd();
+                  director.onCommit(commitPosition);
+                }
               }
             }
           });

--- a/zeebe/broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/SnapshotDirectorPartitionTransitionStepTest.java
+++ b/zeebe/broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/SnapshotDirectorPartitionTransitionStepTest.java
@@ -105,8 +105,10 @@ class SnapshotDirectorPartitionTransitionStepTest {
     if (Role.LEADER.equals(targetRole)) {
       // verify that the last position is read to notify snapshot director
       verify(logstreamReader, times(1)).seekToEnd();
-      assertThat(transitionContext.getSnapshotDirector().getCommitPosition())
-          .isEqualTo(LAST_LOG_POSITION);
+      final ActorFuture<Long> commitPosition =
+          transitionContext.getSnapshotDirector().getCommitPosition();
+      schedulerExtension.workUntilDone();
+      assertThat(commitPosition.join()).isEqualTo(LAST_LOG_POSITION);
     }
   }
 

--- a/zeebe/broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/SnapshotDirectorPartitionTransitionStepTest.java
+++ b/zeebe/broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/SnapshotDirectorPartitionTransitionStepTest.java
@@ -8,8 +8,8 @@
 package io.camunda.zeebe.broker.system.partitions.impl.steps;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
@@ -22,41 +22,54 @@ import io.camunda.zeebe.broker.system.partitions.impl.AsyncSnapshotDirector;
 import io.camunda.zeebe.broker.system.partitions.impl.steps.PartitionTransitionTestArgumentProviders.TransitionsThatShouldCloseService;
 import io.camunda.zeebe.broker.system.partitions.impl.steps.PartitionTransitionTestArgumentProviders.TransitionsThatShouldDoNothing;
 import io.camunda.zeebe.broker.system.partitions.impl.steps.PartitionTransitionTestArgumentProviders.TransitionsThatShouldInstallService;
-import io.camunda.zeebe.scheduler.ActorSchedulingService;
+import io.camunda.zeebe.logstreams.log.LogStream;
+import io.camunda.zeebe.logstreams.log.LogStreamReader;
+import io.camunda.zeebe.scheduler.future.ActorFuture;
+import io.camunda.zeebe.scheduler.testing.ControlledActorSchedulerExtension;
 import io.camunda.zeebe.scheduler.testing.TestActorFuture;
 import io.camunda.zeebe.stream.impl.StreamProcessor;
 import io.camunda.zeebe.util.health.HealthMonitor;
 import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.extension.RegisterExtension;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.ArgumentsSource;
 import org.junit.jupiter.params.provider.EnumSource;
 
 class SnapshotDirectorPartitionTransitionStepTest {
 
+  private static final long LAST_LOG_POSITION = 9;
   TestPartitionTransitionContext transitionContext = new TestPartitionTransitionContext();
+
+  @RegisterExtension
+  private final ControlledActorSchedulerExtension schedulerExtension =
+      new ControlledActorSchedulerExtension();
 
   private SnapshotDirectorPartitionTransitionStep step;
   private final RaftPartition raftPartition = mock(RaftPartition.class);
   private final RaftPartitionServer raftServer = mock(RaftPartitionServer.class);
-  private final ActorSchedulingService actorSchedulingService = mock(ActorSchedulingService.class);
   private final AsyncSnapshotDirector snapshotDirectorFromPrevRole =
       mock(AsyncSnapshotDirector.class);
+  private final LogStream logStream = mock(LogStream.class);
+  private final LogStreamReader logstreamReader = mock(LogStreamReader.class);
 
   @BeforeEach
   void setup() {
     transitionContext.setComponentHealthMonitor(mock(HealthMonitor.class));
     transitionContext.setStreamProcessor(mock(StreamProcessor.class));
     transitionContext.setBrokerCfg(new BrokerCfg());
+    transitionContext.setLogStream(logStream);
 
     when(raftPartition.getServer()).thenReturn(raftServer);
     transitionContext.setRaftPartition(raftPartition);
 
-    when(actorSchedulingService.submitActor(any(), any()))
-        .thenReturn(TestActorFuture.completedFuture(null));
-    transitionContext.setActorSchedulingService(actorSchedulingService);
+    // Use the real ActorScheduler from the extension instead of a mock
+    transitionContext.setActorSchedulingService(schedulerExtension.getActorScheduler());
 
     when(snapshotDirectorFromPrevRole.closeAsync())
         .thenReturn(TestActorFuture.completedFuture(null));
+
+    when(logStream.newLogStreamReader()).thenReturn(logstreamReader);
+    when(logstreamReader.seekToEnd()).thenReturn(LAST_LOG_POSITION);
 
     step = new SnapshotDirectorPartitionTransitionStep();
   }
@@ -89,6 +102,12 @@ class SnapshotDirectorPartitionTransitionStepTest {
     assertThat(transitionContext.getSnapshotDirector())
         .isNotNull()
         .isNotEqualTo(existingSnapshotDirector);
+    if (Role.LEADER.equals(targetRole)) {
+      // verify that the last position is read to notify snapshot director
+      verify(logstreamReader, times(1)).seekToEnd();
+      assertThat(transitionContext.getSnapshotDirector().getCommitPosition())
+          .isEqualTo(LAST_LOG_POSITION);
+    }
   }
 
   @ParameterizedTest
@@ -128,8 +147,12 @@ class SnapshotDirectorPartitionTransitionStepTest {
   }
 
   private void transitionTo(final Role role) {
-    step.prepareTransition(transitionContext, 1, role).join();
-    step.transitionTo(transitionContext, 1, role).join();
+    final ActorFuture<Void> prepareFuture = step.prepareTransition(transitionContext, 1, role);
+    schedulerExtension.workUntilDone();
+    prepareFuture.join();
+    final ActorFuture<Void> transition = step.transitionTo(transitionContext, 1, role);
+    schedulerExtension.workUntilDone();
+    transition.join();
     transitionContext.setCurrentRole(role);
   }
 }

